mautrix/client/api/modules/media_repository.py

```diff
@@ -28,14 +29,65 @@ class MediaRepositoryMethods(BaseClientAPI):
     downloading content from the media repository and for getting URL previews without leaking
     client IPs.
 
-    See also: `API reference <https://matrix.org/docs/spec/client_server/r0.4.0.html#id112>`__"""
+    See also: `API reference <https://matrix.org/docs/spec/client_server/r0.4.0.html#id112>`__
+
+    There are also methods for supporting `MSC2246
+    <https://github.com/matrix-org/matrix-spec-proposals/pull/2246>`__ which allows asynchronous
+    uploads of media.
+    """
+
+    async def create_mxc(self) -> ContentURI:
+        """
+        Create a media ID for uploading media to the homeserver. Requires the homeserver to have
+        `MSC2246 <https://github.com/matrix-org/matrix-spec-proposals/pull/2246>`__ support.
+
+        Returns:
+            The MXC URI that can be used to upload a file to later.
+
+        Raises:
+            MatrixResponseError: If the response does not contain a ``content_uri`` field.
+        """
+        resp = await self.api.request(Method.PUT, MediaPath.unstable["fi.mau.msc2246"].create)
+        try:
+            return resp["content_uri"]
+        except KeyError:
+            raise MatrixResponseError("`content_uri` not in response.")
+
+    async def upload_async(
+        self,
+        data: bytes | bytearray | AsyncIterable[bytes],
+        mime_type: str | None = None,
+        filename: str | None = None,
+        size: int | None = None,
+    ) -> ContentURI:
+        """
+        Create a blank content URI with create_mxc, then start uploading the data in the background
+        and returns the created MXC immediately. Requires the homeserver to have `MSC2246
+        <https://github.com/matrix-org/matrix-spec-proposals/pull/2246>`__ support.
+
+        Args:
+            data: The data to upload.
+            mime_type: The MIME type to send with the upload request.
+            filename: The filename to send with the upload request.
+            size: The file size to send with the upload request.
+
+        Returns:
+            The MXC URI of the file being uploaded asynchronously.
+
+        Raises:
+            MatrixResponseError: If the response does not contain a ``content_uri`` field.
+        """
+        content_uri = await self.create_mxc()
+        asyncio.create_task(self.upload_media(data, mime_type, filename, size, content_uri))
+        return content_uri
```

Comment:
I wonder if the whole `upload_async` method should just be an `async_upload` boolean parameter for `upload_media` :thinking: (doing the `self.api.request` part in a task)